src/Microsoft.AspNetCore.Razor/CodeGenerators/CSharpTagHelperCodeRenderer.cs

```diff
@@ -377,6 +386,36 @@ protected virtual string GenerateUniqueId()
             }
         }
 
+        private void RenderPreallocatedAttributeValues(
+            PreallocatedTagHelperAttributeChunk preallocatedAttributeValueChunk,
+            string valueAccessor)
+        {
+            var attributeValueAccessor = string.Format(
+                    CultureInfo.InvariantCulture,
+                    "{0}.{1}",
+                    preallocatedAttributeValueChunk.AttributeVariableAccessor,
+                    _tagHelperContext.TagHelperAttributeValuePropertyName);
+
+            _writer
+                .WriteStartAssignment(valueAccessor)
+                .WriteInstanceMethodInvocation(
+                    attributeValueAccessor,
+                    nameof(attributeValueAccessor.ToString));
```

Comment:
Perhaps more obvious to unconditionally cast `preallocatedAttribute.Value` to `string`.